lib/mongodb.js

```diff
@@ -2161,47 +2113,76 @@ function visitAllProperties(data, modelCtor, visitor) {
  */
-function coercePropertyValue(modelCtor, propValue, propDef, setValue) {
-  let coercedValue;
-  // Process only mongo-specific data types
-  if (propDef && propDef.mongodb && propDef.mongodb.dataType) {
-    const dataType = propDef.mongodb.dataType;
-    if (typeof dataType === 'string') {
-      if (hasDataType('decimal128', propDef)) {
-        if (Array.isArray(propValue)) {
-          coercedValue = propValue.map(val => Decimal128.fromString(val));
-          return setValue(coercedValue);
-        } else {
-          coercedValue = Decimal128.fromString(propValue);
-          return setValue(coercedValue);
-        }
-      } else if (typeIsObjectId(dataType)) {
-        if (Array.isArray(propValue)) {
-          propValue = propValue.map(val => {
-            if (isObjectIDProperty(modelCtor, propDef, val)) {
-              return coercedValue = ObjectID(val);
-            } else {
-              throw new Error(`${val} is not an ObjectID string`);
-            }
-          });
-          return setValue(propValue);
-        } else if (isObjectIDProperty(modelCtor, propDef, propValue)) {
-          coercedValue = ObjectID(propValue);
-          return setValue(coercedValue);
-        } else {
-          throw new Error(`${propValue} is not an ObjectID string`);
-        }
+function coercePropertyForDatabase(modelCtor, propValue, propDef, setValue) {
+  if (propDef) {
+    let coercedValue;
+    if (hasDataType('decimal128', propDef)) {
+      if (Array.isArray(propValue)) {
+        coercedValue = propValue.map(val => Decimal128.fromString(val));
+        return setValue(coercedValue);
+      } else {
+        coercedValue = Decimal128.fromString(propValue);
+        return setValue(coercedValue);
       }
+    } else {
+      coercedValue = coercePropertyToObjectId(propDef, propValue);
+      return setValue(coercedValue);
     }
   } else {
-    // Object ID coercibility depends on multiple factors, let coerceToObjectId() handle it
+    return setValue(propValue);
+  }
+}
+
+function coercePropertyForUser(modelCtor, propValue, propDef, setValue) {
+  if (propDef) {
+    const uncoercedValue = coercePropertyToString(propDef, propValue);
+    return setValue(uncoercedValue);
+  } else {
+    return setValue(propValue);
+  }
+}
+
+// Coerce a property to ObjectID if it is defined as one
+function coercePropertyToObjectId(propDef, propValue) {
+  if (propValue == null) return propValue;
+  if (hasDataType('objectid', propDef)) {
     if (Array.isArray(propValue)) {
-      propValue = propValue.map(val => coerceToObjectId(modelCtor, propDef, val));
+      propValue = propValue.map(val => ObjectID(val));
+      return propValue;
     } else {
-      propValue = coerceToObjectId(modelCtor, propDef, propValue);
+      return ObjectID(propValue);
+    }
+  } else if (propDef.id) {
+    if (typeof propValue !== 'number') {
+      // Autogenerated id
+      if (propDef.generated) {
+        return ObjectID(propValue);
```

Comment:
Since the id is generated by MongoDB, it will be ObjectId-like string in that block. `ObjectID()` will return the input as it is, if it cannot be converted to `ObjectId`.